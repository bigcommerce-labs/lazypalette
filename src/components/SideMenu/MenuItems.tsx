import { Icon } from 'pattern-lab';
import React, { Component } from 'react';
import { withRouter, RouteComponentProps } from 'react-router-dom';

import ThemeSettings from '../ThemeSettings/ThemeSettings';
import ThemeVariations from '../ThemeVariations/ThemeVariations';

import { activeClassName, NavItem, StyledMenuItems, StyledMenuItemIcon } from './styles';

interface MenuItem {
    disabled?: boolean;
    label: string;
    path: string;
}

interface MenuItemsProps extends RouteComponentProps<{}> {
    items: MenuItem[];
    showArrows?: boolean;
    currentPath: string;
}

const ExpandMenuRoutes = ({ route }: { route: string }) => {
    if (route === 'theme') {
        return <ThemeVariations/>;
    }

    if (route.startsWith('style/')) {
        return <ThemeSettings settingsIndex={parseInt(route.split('/')[1], 10)}/>;
    }

    return null;
};

class MenuItems extends Component<MenuItemsProps, {}> {
    static defaultProps = {
        showArrows: false,
    };

    isPathActive = (path: string) => (
        this.props.location.pathname === `${this.props.currentPath}${path}`
    );

    toggleLink = (path: string) => (
        this.isPathActive(path) ? this.props.currentPath : `${this.props.currentPath}${path}`
    );

    render() {
        return (
            <StyledMenuItems>
                {this.props.items.map(({ disabled, path, label }) => (
                    <li key={path}>
                        <ExpandMenuRoutes route={path}/>
                        <NavItem
                            to={this.toggleLink(path)}
                            exact
                            isActive={(match, location) => this.isPathActive(path)}
                            activeClassName={activeClassName}
                            disabled={disabled}>
                            {label}
                            {this.props.showArrows &&
                                <StyledMenuItemIcon>
                                    <Icon glyph="chevronRight" size="large" />
                                </StyledMenuItemIcon>
                            }
                        </NavItem>
                    </li>
                ))}
            </StyledMenuItems>
        );
    }
}

export default withRouter(MenuItems);
