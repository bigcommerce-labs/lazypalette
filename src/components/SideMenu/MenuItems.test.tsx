import { shallow } from 'enzyme';
import React from 'react';

import MenuItems from './MenuItems';

it('renders', () => {
  const testItems = [
    {
      label: 'Logo and Store Name',
      path: 'logo',
    },
    {
      label: 'Store Theme',
      path: 'theme',
    },
  ];

  const menuItems = shallow(
    <MenuItems items={testItems} currentPath="/"/>
  );

  expect(menuItems).toMatchSnapshot();
});
