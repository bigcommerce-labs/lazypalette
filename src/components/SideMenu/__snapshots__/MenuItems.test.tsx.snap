// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MenuItems
    currentPath="/"
    items={
      Array [
        Object {
          "label": "Logo and Store Name",
          "path": "logo",
        },
        Object {
          "label": "Store Theme",
          "path": "theme",
        },
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <li>
          <Styled(NavLink)
            activeClassName="nav-item-active"
            exact={true}
            to="/logo"
          >
            Logo and Store Name
          </Styled(NavLink)>
        </li>,
        <li>
          <Styled(NavLink)
            activeClassName="nav-item-active"
            exact={true}
            to="/theme"
          >
            Store Theme
          </Styled(NavLink)>
        </li>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "logo",
        "nodeType": "host",
        "props": Object {
          "children": <Styled(NavLink)
            activeClassName="nav-item-active"
            exact={true}
            to="/logo"
          >
            Logo and Store Name
          </Styled(NavLink)>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "activeClassName": "nav-item-active",
            "children": "Logo and Store Name",
            "exact": true,
            "to": "/logo",
          },
          "ref": null,
          "rendered": "Logo and Store Name",
          "type": [Function],
        },
        "type": "li",
      },
      Object {
        "instance": null,
        "key": "theme",
        "nodeType": "host",
        "props": Object {
          "children": <Styled(NavLink)
            activeClassName="nav-item-active"
            exact={true}
            to="/theme"
          >
            Store Theme
          </Styled(NavLink)>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "activeClassName": "nav-item-active",
            "children": "Store Theme",
            "exact": true,
            "to": "/theme",
          },
          "ref": null,
          "rendered": "Store Theme",
          "type": [Function],
        },
        "type": "li",
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <li>
            <Styled(NavLink)
              activeClassName="nav-item-active"
              exact={true}
              to="/logo"
            >
              Logo and Store Name
            </Styled(NavLink)>
          </li>,
          <li>
            <Styled(NavLink)
              activeClassName="nav-item-active"
              exact={true}
              to="/theme"
            >
              Store Theme
            </Styled(NavLink)>
          </li>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "logo",
          "nodeType": "host",
          "props": Object {
            "children": <Styled(NavLink)
              activeClassName="nav-item-active"
              exact={true}
              to="/logo"
            >
              Logo and Store Name
            </Styled(NavLink)>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "activeClassName": "nav-item-active",
              "children": "Logo and Store Name",
              "exact": true,
              "to": "/logo",
            },
            "ref": null,
            "rendered": "Logo and Store Name",
            "type": [Function],
          },
          "type": "li",
        },
        Object {
          "instance": null,
          "key": "theme",
          "nodeType": "host",
          "props": Object {
            "children": <Styled(NavLink)
              activeClassName="nav-item-active"
              exact={true}
              to="/theme"
            >
              Store Theme
            </Styled(NavLink)>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "activeClassName": "nav-item-active",
              "children": "Store Theme",
              "exact": true,
              "to": "/theme",
            },
            "ref": null,
            "rendered": "Store Theme",
            "type": [Function],
          },
          "type": "li",
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
